java/azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/EventData.java

```diff
@@ -196,43 +207,130 @@ public EventData(ByteBuffer buffer)
+	{
+		return this.bodyData == null ? null : this.bodyData.getArray();
+	}
+	
+	/**
+	 * Get Actual Payload/Data wrapped by EventData.
+	 * This is intended to be used after receiving EventData using @@PartitionReceiver.
+	 * @return returns the byte[] of the actual data
+	 * @see #getPayloadArrayLength()
+	 * @see #getPayloadArray()
+	 */
+	public int getPayloadArrayOffset()
+	{
+		return this.bodyData == null ? 0 : this.bodyData.getArrayOffset();
+	}
+	
+	/**
+	 * Get Actual Payload/Data wrapped by EventData.
+	 * This is intended to be used after receiving EventData using @@PartitionReceiver.
+	 * @return returns the byte[] of the actual data
+	 * @see #getPayloadArray()
+	 * @see #getPayloadArrayOffset()
+	 */
+	public int getPayloadArrayLength()
+	{
+		return this.bodyData == null ? 0 : this.bodyData.getLength();
+	}
 
 	/**
 	 * Application property bag
 	 * @return returns Application properties
 	 */
 	public Map<String, String> getProperties()
 	{
-		return this.properties;
-	}
+		if (this.properties == null)
+		{
+			this.properties = new HashMap<String, String>();
+		}
 
-	public void setProperties(Map<String, String> applicationProperties)
-	{
-		this.properties = applicationProperties;
+		return this.properties;
 	}
 
 	/**
 	 * SystemProperties that are populated by EventHubService.
 	 * <p>As these are populated by Service, they are only present on a Received EventData.
 	 * @return an encapsulation of all SystemProperties appended by EventHubs service into EventData
+	 * @deprecated As of release 0.8.0. Use {@link #getSystemProperties()}
 	 */
-	public SystemProperties getSystemProperties()
+	@Deprecated public SystemProperties getDeprecatedSystemProperties()
 	{
-		if (this.isReceivedEvent && this.systemProperties == null)
+		if (this.isReceivedEvent && this.deprecatedSystemProperties == null)
 		{
-			this.systemProperties = new SystemProperties(this);
+			this.deprecatedSystemProperties = new SystemProperties(this);
 		}
 
+		return this.deprecatedSystemProperties;
+	}
+	
+	/**
+	 * {@link #getSystemProperties()} from version 0.7.9 was deprecated and renamed to {@link #getDeprecatedSystemProperties()} starting 0.8.0.
+	 * {@link #getOffset()}, {@link #getSequenceNumber()}, {@link #getPartitionKey()}, {@link #getEnqueuedTime()} are now available directly on {@link EventData} class.
+	 * <p>SystemProperties that are populated by EventHubService.
+	 * <p>As these are populated by Service, they are only present on a Received EventData.
+	 * <p>Usage:<p>
+	 * <code>
+	 * final String offset = eventData.getSystemProperties().get(AmqpConstants.OFFSET_ANNOTATION_NAME);
+	 * </code>
+	 * @return an encapsulation of all SystemProperties appended by EventHubs service into EventData.
+	 * <code>null</code> if the {@link #EventData()} is not received and is created by the public constructors. 
+	 */
+	public Map<String, Object> getSystemProperties()
+	{
 		return this.systemProperties;
 	}
-
+	
+	public String getOffset()
```

Comment:
>getOffset [](start = 15, length = 9)

I would say we should not have these property method in EventData because these are meaningless in send scenario, but we don't separate SendEvetData/ReceivedEventData so these methods "in theory" should check if the data is a received data.

Rather than doing the check, user can simply get to the system properties and get the data from there themselves.